subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseWtpFacet.java

```diff
@@ -138,17 +140,26 @@ public void setFacets(List<Facet> facets) {
      * <p>
      * If a facet already exists with the given name then its version will be updated.
      * <p>
+     * In the case of a "jst.ejb" facet, the incompatible "jst.utility" installed by default is also removed.
+     * </p>
      * For examples see docs for {@link EclipseWtpFacet}
      *
      * @param args A map that must contain a 'name' and 'version' key with corresponding values.
      */
     public void facet(Map<String, ?> args) {
         Facet newFacet = ConfigureUtil.configureByMap(args, new Facet());
+        List<Facet> newFacets;
```

Comment:
❌ If the client sets the facets with `setFacets()` then the facet replacement check would not kick in. One could argue that then it's the clients responsibility there, at the same time, it the inconsistency between those methods can be very confusing.

A better place for the facet replacement would be `EclipseWtpFacet.mergeXmlFacet()`. It would be always called and the client would still have the chance to do final adjustments in the `whenMerged` callback.

```java
    public void mergeXmlFacet(WtpFacet xmlFacet) {
        file.getBeforeMerged().execute(xmlFacet);
        xmlFacet.configure(replaceInconsistentFacets(getFacets()));
        file.getWhenMerged().execute(xmlFacet);
    }
}
```

Reply:
If the verification should be done when the client sets the facets with `setFacets`, we could also put it there: 

```java
    public void setFacets(List<Facet> facets) {
        this.facets = replaceInconsistentFacets(facets);
    }
```

Reply:
If the replacement should be done when client sets the facets with `setFacets()`, why not put it in the setter ?

```java
    public void setFacets(List<Facet> facets) {
        this.facets = replaceInconsistentFacets(facets);
    }
```

Reply:
It's mostly a nitpick from my side, but calling `replaceInconsistentFacets()` in `mergeXmlFacet()` ensures that the filtering/replacement happens exactly once. If you see any downsides then I'm fine with moving it `facets()` and `setFacets()`.